### PR TITLE
Improve configuration dumping logic

### LINKED ISSUE #1915: get-config output should be a valid suite definition
(apparently it doesn't at the moment!)

### PR DESCRIPTION
**Pull request by matthewrmshin**
The configuration generated by `cylc get-config` and in a suite's `~/cylc-run/SUITE/log/suiterc/` should now pass `cylc validate`.

Ability to skip printing settings with `None` values.
Duration will now be dumped in ISO8601 format by default.
Improve quoting of values in lists.
Triple quote multi-line values.
Allow empty values for settings expecting a set of predefined strings.
Settings expecting `int`, `float` and `boolean` are now set to `None` if given empty values.
If an `inherit` setting has an empty value, it will default to `inherit = root`.
4 spaces indent instead of 3.

And some style updates to pass pylint and pep8.

Close #1915.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
(`parsec.util` and `parsec.validate` will now pass `pep8` in my environment.)

**Comment by matthewrmshin** [comment]
@hjoliver @arjclark please review, sanity check or reassign.

File: lib/cylc/cfgspec/suite.py [hunk 1]
```diff
@@ -428,6 +432,7 @@ def _coerce_final_cycletime(value, keys, args):
 
 
 def upg(cfg, descr):
+    """Upgrader."""
```
Comment (arjclark) [inline lib/cylc/cfgspec/suite.py]: More "upgrade syntax" rather than "upgrader" I'd say

File: lib/cylc/syntax_flags.py [hunk 2]
```diff
@@ -43,26 +43,14 @@ def __str__(self):
...
-                       exc_args=None, exc_kwargs=None):
+def set_syntax_version(version, message):
     """Attempt to define the syntax version.
 
-    If it's already defined differently, raise an exception.
-    The exception is exc_class if it is not None, raised
-    using *exc_args and **exc_kwargs (or using message as the
-    first argument if exc_args and exc_kwargs aren't set).
-    Otherwise, raise SyntaxVersionError.
+    If it's already defined differently, raise SyntaxVersionError.
```
Comment (arjclark) [inline lib/cylc/syntax_flags.py]: Probably get rid of this blank line

**Comment by arjclark** [comment]
Review 2 ok.

**Comment by arjclark** [comment]
@matthewrmshin - branch now in conflict

**Comment by matthewrmshin** [comment]
Branch re-based. Conflicts resolved. Tests OK (with and without global configuration) in my environment.

**Comment by matthewrmshin** [comment]
Branch re-based. Tests OK (with and without global configuration) in my environment.

File: lib/parsec/util.py [hunk 3]
```diff
@@ -15,123 +15,140 @@
...
-                # delay output to print top level items before recursing
-                delayed.append((key,val))
-            else:
-                # val is a single value
-                if isinstance( val, list ):
-                    v = listjoin( val, none_str )
-                elif val is None:
-                    v = none_str
+    stack = [("", cfg, level, indent)]
+    while stack:
```
Comment (hjoliver) [inline lib/parsec/util.py]: So you don't like recursion...?

File: lib/parsec/util.py [hunk 4]
```diff
@@ -15,123 +15,140 @@
...
     """
     if not source:
         target = OrderedDictWithDefaults()
         return
     if hasattr(source, "defaults_"):
         target.defaults_ = pdeepcopy(source.defaults_)
-    for key,val in source.items():
-        if isinstance( val, dict ):
+    for key, val in source.items():
+        if isinstance(val, dict):
```
Comment (hjoliver) [inline lib/parsec/util.py]: I guess we're not applying PEP8 to `lib/parsec`.  OK, we sure aren't (just checked).  My fault.

**Comment by hjoliver** [comment]
Review 1 - no problems found.  The pep8 test is failing on `lib/cylc/gui/updater_dot.py` but the problem is fixed on current master.

**Pull request merged**